compiler/rustc_borrowck/src/region_infer/mod.rs

```diff
@@ -1185,33 +1211,71 @@ impl<'tcx> RegionInferenceContext<'tcx> {
             }
 
             VerifyBound::AnyBound(verify_bounds) => verify_bounds.iter().any(|verify_bound| {
-                self.eval_verify_bound(tcx, body, generic_ty, lower_bound, verify_bound)
+                self.eval_verify_bound(
+                    infcx,
+                    param_env,
+                    body,
+                    generic_ty,
+                    lower_bound,
+                    verify_bound,
+                )
             }),
 
             VerifyBound::AllBounds(verify_bounds) => verify_bounds.iter().all(|verify_bound| {
-                self.eval_verify_bound(tcx, body, generic_ty, lower_bound, verify_bound)
+                self.eval_verify_bound(
+                    infcx,
+                    param_env,
+                    body,
+                    generic_ty,
+                    lower_bound,
+                    verify_bound,
+                )
             }),
         }
     }
 
     fn eval_if_eq(
         &self,
-        tcx: TyCtxt<'tcx>,
-        body: &Body<'tcx>,
+        infcx: &InferCtxt<'_, 'tcx>,
         generic_ty: Ty<'tcx>,
         lower_bound: RegionVid,
         test_ty: Ty<'tcx>,
-        verify_bound: &VerifyBound<'tcx>,
+        verify_bound: Region<'tcx>,
     ) -> bool {
-        let generic_ty_normalized = self.normalize_to_scc_representatives(tcx, generic_ty);
-        let test_ty_normalized = self.normalize_to_scc_representatives(tcx, test_ty);
+        let generic_ty_normalized = self.normalize_to_scc_representatives(infcx.tcx, generic_ty);
+        let test_ty_normalized = self.normalize_to_scc_representatives(infcx.tcx, test_ty);
         if generic_ty_normalized == test_ty_normalized {
-            self.eval_verify_bound(tcx, body, generic_ty, lower_bound, verify_bound)
+            let verify_bound_vid = self.to_region_vid(verify_bound);
+            self.eval_outlives(verify_bound_vid, lower_bound)
         } else {
             false
         }
     }
 
+    fn eval_if_eq_bound(
```

Comment:
Is the separation of bound/not bound solely a performance thing? Could we merge them and make the decision on whether the binder binds nothing?

Reply:
@oli-obk I'm removing the "bound" variant, I thought I pushed that but I guess not.

Reply:
er, removing the *not* bound variant. it's dead code in this PR anyway.